nano-sync/src/circuits/mnt4/pk_tree_5.rs

```diff
@@ -0,0 +1,314 @@
+            cs.ns(|| "alloc pedersen_generators"),
+            pedersen_generators(195),
+        )?;
+
+        // Allocate all the private inputs.
+        next_cost_analysis!(cs, cost, || { "Alloc private inputs" });
+
+        let pks_var = Vec::<G2Gadget>::alloc(cs.ns(|| "alloc public keys"), || Ok(&self.pks[..]))?;
+
+        let pks_nodes_var =
+            Vec::<G1Gadget>::alloc(cs.ns(|| "alloc pks Merkle proof nodes"), || {
+                Ok(&self.pks_nodes[..])
+            })?;
+
+        let prepare_agg_pk_var =
+            G2Gadget::alloc(cs.ns(|| "alloc prepare aggregate public key"), || {
+                Ok(&self.prepare_agg_pk)
+            })?;
+
+        let commit_agg_pk_var =
+            G2Gadget::alloc(cs.ns(|| "alloc commit aggregate public key"), || {
+                Ok(&self.commit_agg_pk)
+            })?;
+
+        // Allocate all the public inputs.
+        next_cost_analysis!(cs, cost, || { "Alloc public inputs" });
+
+        let pk_commitment_var = UInt8::alloc_input_vec(
+            cs.ns(|| "alloc public keys commitment"),
+            self.pks_commitment.as_ref(),
+        )?;
+
+        let prepare_signer_bitmap_var = UInt8::alloc_input_vec(
+            cs.ns(|| "alloc prepare signer bitmap"),
+            self.prepare_signer_bitmap.as_ref(),
+        )?;
+
+        let prepare_agg_pk_commitment_var = UInt8::alloc_input_vec(
+            cs.ns(|| "alloc prepare aggregate pk commitment"),
+            self.prepare_agg_pk_commitment.as_ref(),
+        )?;
+
+        let commit_signer_bitmap_var = UInt8::alloc_input_vec(
+            cs.ns(|| "alloc commit signer bitmap"),
+            self.commit_signer_bitmap.as_ref(),
+        )?;
+
+        let commit_agg_pk_commitment_var = UInt8::alloc_input_vec(
+            cs.ns(|| "alloc commit aggregate pk commitment"),
+            self.commit_agg_pk_commitment.as_ref(),
+        )?;
+
+        let position_var =
+            UInt8::alloc_input_vec(cs.ns(|| "alloc position"), self.position.as_ref())?
+                .pop()
+                .unwrap();
+
+        // Process public inputs.
+        next_cost_analysis!(cs, cost, || { "Process public inputs" });
+
+        let chunk_start = VALIDATOR_SLOTS / PK_TREE_BREADTH * (self.position[0] as usize);
+        let chunk_end = VALIDATOR_SLOTS / PK_TREE_BREADTH * (1 + self.position[0] as usize);
+
+        let prepare_signer_bitmap_var =
+            prepare_signer_bitmap_var.to_bits(cs.ns(|| "prepare signer bitmap to bits"))?;
+
+        let prepare_signer_bitmap_var = prepare_signer_bitmap_var[chunk_start..chunk_end].to_vec();
+
+        let commit_signer_bitmap_var =
+            commit_signer_bitmap_var.to_bits(cs.ns(|| "commit signer bitmap to bits"))?;
+
+        let commit_signer_bitmap_var = commit_signer_bitmap_var[chunk_start..chunk_end].to_vec();
+
+        let mut path_var = position_var.into_bits_le();
+
+        path_var.truncate(PK_TREE_DEPTH);
```

Comment:
It's non-obvious what's happening here. I would add a comment.

Reply:
Agree, I'll add some explanation.

Reply:
Done.